File: 2016/12.py
with open("input_12.txt") as f:
    lines = f.readlines()

for i in range(len(lines)):
    lines[i] = lines[i].split("\n")[0]

part = 2
if part == 1:    
    registry = {"a":0, "b":0, "c":0, "d":0}
else:
    registry = {"a":0, "b":0, "c":1, "d":0}

line = 0
while line < len(lines):
    #print (lines[line])
    parts = lines[line].split()
    cmd = parts[0]
    params = parts[1:]
    jump = False
    
    if cmd == "cpy":
        try:
            x = int(params[0])
        except ValueError:
            x = registry[params[0]]
        registry[params[1]] = x
    elif cmd == "inc":
        registry[params[0]] += 1
    elif cmd == "dec":
        registry[params[0]] -= 1
    elif cmd == "jnz":
        try:
            x = int(params[0])        
        except ValueError:
            x = registry[params[0]]
        if x != 0:
            jump = True
            line += int(params[1])
    if not jump:
        line += 1

if part == 1:
    print ("🎄Part 1: {}".format(registry['a']))
else:
    print("🎁🎄Part 2: {}".format(registry['a']))

File: 2016/12.jl
include("utils.jl")
using .AoCUtils

register = Dict("a"=>0, "b"=>0, "c"=>0, "d"=>0)

function check_str2(a)
    return tryparse(Float64, a) !== nothing
end

function parseArg(arg)
    if check_str2(arg)
        return parse(Int, arg)
    else
        return register[arg]
    end
end

function compiler(lines)
    cline = 1
    while true
        l = lines[cline]
        if startswith(l, "cpy")
            args = split(l, " ")[2:end]
            register[args[2]] = parseArg(args[1])
            cline += 1
        elseif startswith(l, "inc")
            arg = split(l, " ")[end]
            register[arg] += 1
            cline += 1
        elseif startswith(l, "dec")
            arg = split(l, " ")[end]
            register[arg] -= 1
            cline += 1
        elseif startswith(l, "jnz")
            args = split(l, " ")[2:end]
            if parseArg(args[1]) != 0
                cline += parse(Int, args[2])
            else
                cline += 1
            end
        end
        #println(cline)
        if cline < 1 || cline >length(lines)
            break
        end
    end

    println("🎅 $(register["a"])")
end

lines = readLines("input_12.txt")

compiler(lines)
register["c"] = 1
compiler(lines)
